keps/sig-node/0036-20190118-ForceEviction.md

```diff
@@ -0,0 +1,178 @@
+without waiting for Kubelet to responde. One important use case of this taint is node shut down case. If node
+controller or some external controller detact that node is powered off, the controller could add the following
+taint to the node.
+
+
+```
+taint := &v1.Taint{
+	Key:    nodeShutdown,
+	Value:  "true",
+	Effect: v1.TaintEffectForceEviction,
+}
+```
+
+After the node is tainted, node lifecycle controller will trigger to forcefully delete all the Pods from the node.
+If Kubelet comes back healthy in the middle of force eviction process, there might be race condition between kubelet
+starting Pods while they are being deleted since force deletion does not require kubelet coordination any more.
+To avoid race condition between controller and kubelet, kubelet also requires some additional handling when node
+is tainted which is explained in more details below.
+
+For node lifecycle controller,  we plan to add the following eviction logic:
+
+When node is marked as “NotReady” and tainted with “ForceEviction”. Taint manager will perform the following steps
+
+1. Check node condition whether Ready becomes true again. If yes, remove the taint if ForceEviction pods on the node
+is not started yet or already finished. Otherwise, skip the rest. If Ready is still false, continue with next step 2.
+2. Remove VolumeInUse list for this node and Remove VolumesAttached list
+3. Evict pods from the node with --force and 0 grace period
+
+For node Kubelet, it will first send out heartbeat by updating node status during start up.
+
+Before starting up Pod, Pod canAdmit function checks taint and pod tolerations to decide whether to admit pod or not.
+- If node condition can not be retrieved, do not start Pods that cannot tolerant ForceEviction taint. But we should allow
+to start static Pods.
+- If node has taint or node Ready condition is not true (heartbeat has not yet sent out), do not start Pods
+that cannot tolerant ForceEviction taint
+
+For Daemonsets we should not add the corresponding toleration on the Daemonset controller, as we will end up filling our
+code base with well known Kubernetes taints. Instead we should clearly state on the
+release note the action required for cluster-critical Daemonsets pods.
+
+
+## User Stories
+
+As a kubernetes administrator/IT automation tool, I want workload can be recovered quickly when some nodes fail.
+Some controller could detect the node failures, and could shut it down so that the workload running on the node could
+be recoved on other healthy nodes.
+
+## Alternatives
+
+Kubernetes administrator/IT automation tool could choose to delete the node API object from API server if it detect some
+node failures. In this way, Pods will be garbage collected after a while. However, this way it will lose some meta data that
+were applied to the node already which is not ideal. Also after fixing the node, administroator has to register this node back.
```

Comment:
They can also simply force delete the pods associated with the node

Reply:
The goal here is to avoid burden operations with this, especially when it’s 24/7 run